sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -328,27 +328,46 @@ object TypeCoercion {
    */
   object WidenSetOperationTypes extends Rule[LogicalPlan] {
 
-    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperatorsUp {
-      case s @ Except(left, right, isAll) if s.childrenResolved &&
-        left.output.length == right.output.length && !s.resolved =>
-        val newChildren: Seq[LogicalPlan] = buildNewChildrenWithWiderTypes(left :: right :: Nil)
-        assert(newChildren.length == 2)
-        Except(newChildren.head, newChildren.last, isAll)
-
-      case s @ Intersect(left, right, isAll) if s.childrenResolved &&
-        left.output.length == right.output.length && !s.resolved =>
-        val newChildren: Seq[LogicalPlan] = buildNewChildrenWithWiderTypes(left :: right :: Nil)
-        assert(newChildren.length == 2)
-        Intersect(newChildren.head, newChildren.last, isAll)
-
-      case s: Union if s.childrenResolved && !s.byName &&
+    def apply(plan: LogicalPlan): LogicalPlan = {
+      val exprIdMapArray = mutable.ArrayBuffer[(ExprId, Attribute)]()
+      val newPlan = plan resolveOperatorsUp {
+        case s @ Except(left, right, isAll) if s.childrenResolved &&
+          left.output.length == right.output.length && !s.resolved =>
+          val (newChildren, newExprIds) = buildNewChildrenWithWiderTypes(left :: right :: Nil)
+          exprIdMapArray ++= newExprIds
+          assert(newChildren.length == 2)
+          Except(newChildren.head, newChildren.last, isAll)
+
+        case s @ Intersect(left, right, isAll) if s.childrenResolved &&
+          left.output.length == right.output.length && !s.resolved =>
+          val (newChildren, newExprIds) = buildNewChildrenWithWiderTypes(left :: right :: Nil)
+          exprIdMapArray ++= newExprIds
+          assert(newChildren.length == 2)
+          Intersect(newChildren.head, newChildren.last, isAll)
+
+        case s: Union if s.childrenResolved && !s.byName &&
           s.children.forall(_.output.length == s.children.head.output.length) && !s.resolved =>
-        val newChildren: Seq[LogicalPlan] = buildNewChildrenWithWiderTypes(s.children)
-        s.copy(children = newChildren)
+          val (newChildren, newExprIds) = buildNewChildrenWithWiderTypes(s.children)
+          exprIdMapArray ++= newExprIds
+          s.copy(children = newChildren)
+      }
+
+      // Re-maps existing references to the new ones (exprId and dataType)
+      // for aliases added when widening columns' data types.
```

Comment:
Another common way to solve this issue is to create an `Alias` with the existing exprId, so that we don't need to rewrite the parent nodes.

I think it's safer than rewriting the parent nodes. We rewrite parent nodes in `ResolveReferences.dedupRight`, we hit bugs and end up with a complicated solution.

Reply:
I thought about it too. But I'm not sure if duplicate exprId is okay. If this is common way, it sounds simple and safe.

Reply:
You meant re-alias with exprId=1 in the example above like this?
```
org.apache.spark.sql.AnalysisException: Resolved attribute(s) v#1 missing from v#3 in operator !Project [v#1]. Attribute(s) with the same name appear in the operation: v. Please check if the right attribute(s) are used.;;
!Project [v#1]  <------ the reference got missing
+- SubqueryAlias t
   +- Union
      :- Project [cast(v#1 as decimal(11,0)) AS v#3] <----- !!!! re-alias with exprId=#1 ?!!!!!
      :  +- Project [v#1]
      :     +- SubqueryAlias t3
      :        +- SubqueryAlias tbl
      :           +- LocalRelation [v#1]
      +- Project [v#2]
         +- Project [CheckOverflow((promote_precision(cast(v#1 as decimal(11,0))) + promote_precision(cast(v#1 as decimal(11,0)))), DecimalType(11,0), true) AS v#2]
            +- SubqueryAlias t3
               +- SubqueryAlias tbl
                  +- LocalRelation [v#1]
```

Reply:
Yes, like re-alias with exprId=1

Just did a quick search, rule `TimeWindowing`, `Aggregation` did it. AFAIK it's common when need to change the plan in the middle and don't want to affect the parent nodes.

Reply:
Yea, I tried it first, but `RemoveNoopOperators` will remove a `Project` with a rewritten alias https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L480
Because it assumes projects having common exprIds have the same output. There may be a way to avoid the case and I'll check `TimeWindowing`.

Reply:
In case of `TimeWindowing`, it seems `RemoveNoopOperators` cannot remove a project having the same exprIds because the output numbers are different before/after transformation (L3622).
https://github.com/apache/spark/blob/11c6a23c13745a61c1a1cfc82e4f1ac95eaaa04a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala#L3610-L3623

Looked around the related code though, I couldn't find a solution to avoid the case. Any suggestion?

Reply:
Doesn't `RemoveNoopOperators` check if the outputs are semantically equal? Is `cast(v#1 as decimal(11,0)) AS v#1` semantically equal to `v#1`? `Canonicalize` should keep `cast` and `alias`.

Reply:
hm I see. I'll check that approach.

Reply:
On second thought, we still need to update parent nodes. For example, in the example of the PR description;
```
!Project [v#1]  <------ this project already has `AttributeReference(v, decimal(10, 0))#1`, so
                        we need to update the data type, too
+- SubqueryAlias t
   +- Union
      :- Project [cast(v#1 as decimal(11,0)) AS v#1] <----- re-alias with exprId=#1
      :  +- Project [v#1] <----- dataType=decimal(10, 0)
      :     +- SubqueryAlias t3
      :        +- SubqueryAlias tbl
      :           +- LocalRelation [v#1]

      +- Project [v#2] <----- dataType=decimal(11, 0)
         +- ...
```
the parent `Project` has a attribute reference with `exprId=1` and `dataType=decimal(10, 0)`. So, we need to update the data type, too. If we don't update it, plan integrity can break, e.g., in `PushProjectionThroughUnion`.
```
-- !query
CREATE OR REPLACE TEMPORARY VIEW t3 AS VALUES (decimal(1)) tbl(v)
-- !query schema
struct<>
-- !query output

-- !query
SELECT t.v FROM (
  SELECT v FROM t3
  UNION ALL
  SELECT v + v AS v FROM t3
) t
-- !query schema
struct<>
-- !query output
org.apache.spark.sql.catalyst.errors.package$TreeNodeException
After applying rule org.apache.spark.sql.catalyst.optimizer.PushProjectionThroughUnion in batch Operator Optimization before Inferring Filters, the structural integrity of the plan is broken., tree:
'Union false, false
:- Project [v#183]
:  +- Project [cast(v#183 as decimal(11,0)) AS v#183]
:     +- Project [v#183]
:        +- LocalRelation [v#183]
+- Project [v#184]
   +- Project [v#184]
      +- Project [CheckOverflow((promote_precision(cast(v#183 as decimal(11,0))) + promote_precision(cast(v#183 as decimal(11,0)))), DecimalType(11,0), true) AS v#184]
         +- LocalRelation [v#183]
```

Reply:
This is weird. For all the type coercion rules, they extend `TypeCoercionRule`, which will update parent nodes with data type change automatically.